### PR TITLE
Io log rewrite

### PR DESCRIPTION
**Pull request by vibhavp**
TODO:

- [x] Address  `lsp-log-max`

### CODE REVIEW

File: README.org [hunk 1]
```diff
@@ -152,10 +152,8 @@
    - ~lsp-lens-show~ - Show lenses in the current file
    - ~lsp-lens-hide~ - Hide lenses in the current file
    - ~lsp-lens-mode~ (experimental) - Turn on/off lenses in the current file.
-   - ~lsp-save-logs~ - Save a trace of all messages to and from the language server to a given file. Only works when ~lsp-save-logs~ is non-nil.
```
Comment (yyoncho) [inline README.org]: please update throubleshooting section, add documentation for the new variables, modes, new shortcuts, etc. Update also the bug template.

File: lsp-mode.el [hunk 2]
```diff
@@ -1887,52 +1882,67 @@ TYPE can either be 'incoming or 'outgoing"
...
-                     method id process-time))))
-                (insert "\n")
-                (insert (if (memq type '(incoming-resp ougoing-resp))
-                            "Result: "
-                          "Params: "))
-                (insert (json-encode (lsp--log-entry-body entry)))
-                (insert "\n\n\n")))
-            trace)
-      (write-region nil nil file))))
+(define-inline lsp--log-entry-new (entry workspace)
```
Comment (yyoncho) [inline lsp-mode.el]: as per offline discussion better to use defun's by default and optimize only after doing the proper performance testing.

File: lsp-mode.el [hunk 3]
```diff
@@ -4633,6 +4644,20 @@ SESSION is the active session."
       (lsp--start-workspace session client project-root (lsp--create-initialization-options session client))
     (lsp--spinner-stop)))
 
+(defun lsp--log-io-next (arg)
```
Comment (yyoncho) [inline lsp-mode.el]: this should be public (only one `-`)

File: lsp-mode.el [hunk 4]
```diff
@@ -4633,6 +4644,20 @@ SESSION is the active session."
...
+  (interactive "P")
+  (ewoc-goto-next lsp--log-io-ewoc (or arg 1)))
+
+(defun lsp--log-io-prev (arg)
```
Comment (yyoncho) [inline lsp-mode.el]: ditto

File: lsp-mode.el [hunk 5]
```diff
@@ -4633,6 +4644,20 @@ SESSION is the active session."
...
+  (interactive "P")
+  (ewoc-goto-prev lsp--log-io-ewoc (or arg 1)))
+
+(define-derived-mode lsp-log-io-mode view-mode "LspLogIo"
+  "Special mode for viewing IO logs.")
+
+(define-key lsp-log-io-mode-map (kbd "<down>") #'lsp--log-io-next)
```
Comment (yyoncho) [inline lsp-mode.el]: Can you move this to M-n/M-p? These are more 'emacsy'.

Comment (vibhavp) [reply lsp-mode.el]: Done

**Comment by yyoncho** [comment]
@vibhavp can you fix conflicts?

**Pull request merged**